File: summer2018/easy/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        map<char, char> m = {
            {'}', '{'}, {']', '['}, {')', '('}
        };
        stack<char> stk;
        for(int i = 0; i < s.length(); ++i){
            if(s[i] == '}' || s[i] == ']' || s[i] == ')'){
                if(stk.empty() || stk.top() != m[s[i]])
                    return false;
                else
                    stk.pop();
            } else 
                stk.push(s[i]);
        }
        return stk.empty();
    }
};

File: language/python/20-valid-parentheses.py
# My method
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        left = "({["
        pair = {')':'(', ']':'[', '}':'{'}
        stack = [] 
        for c in s:
            if c in left:
                stack.insert(0, c)
            else:    
                if stack and pair[c] == stack[0]:
                    stack.pop(0)
                else:
                    return False 
        return not stack
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        pair = {')':'(', ']':'[', '}':'{'}
        stack = [] 
        for c in s:
            if c in pair.values():
                stack.append(c)
            else:    
                if stack and pair[c] == stack[-1]:
                    stack.pop()
                else:
                    return False 
        return not stack
# Better method
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        amap = {
            "(": ")",
            "{": "}",
            "[": "]"
        }
        stack = []
        for c in s:
            if c in amap:
                stack.append(amap[c])
            else:
                if not stack:
                    return False
                val = stack.pop()
                if c != val:
                    return False
        return len(stack) == 0
